tests/run/i18315.scala

```diff
@@ -0,0 +1,15 @@
+trait Sam1:
+  type T
+  def apply(x: T): T
+
+trait Sam2:
+  var x: Int = 1 // To force anonymous class generation
+  type T
+  def apply(x: T): T
+
+object Test:
+  def main(args: Array[String]): Unit =
+    val s1: Sam1 { type T = String } = x => x.trim
```

Comment:
We should definitely not make *more* things work in this space. What this PR "fixes" is already off-spec, because it is not legal to put a refinement type in an `extends` clause. It happened to work in 3.3.0 by accident, and so we're preserving the status quo in the name of backward compatibility within a patch series. But we should in fact make it an error in 3.4.x.

Reply:
Yeah, I wouldn't want to add more than this and definitely not in a PR that fixes a regression